File: product-of-array-except-self/tedkimdev.go
// TC: O(n)
// SC: O(n)
func productExceptSelf(nums []int) []int {
	result := make([]int, 0)

	left := make([]int, len(nums))
	left[0] = 1

	right := make([]int, len(nums))
	right[len(nums)-1] = 1

	for i := 1; i < len(nums); i++ {
		left[i] = nums[i-1] * left[i-1]
	}

	for i := len(nums) - 2; i >= 0; i-- {
		right[i] = nums[i+1] * right[i+1]
	}

	for i := 0; i < len(nums); i++ {
		result = append(result, left[i]*right[i])
	}
	return result
}

File: spiral-matrix/tedkimdev.rs
// TC: O(m * n)
// SC: O(m * n)
impl Solution {
    pub fn spiral_order(matrix: Vec<Vec<i32>>) -> Vec<i32> {
        let mut result: Vec<i32> = Vec::new();

        let mut left = 0i32;
        let mut right = matrix[0].len() as i32;
        let mut top = 0i32;
        let mut bottom = matrix.len() as i32;

        while left < right && top < bottom {
            for i in left..right {
                result.push(matrix[top as usize][i as usize]);
            }
            top += 1;
            for i in top..bottom {
                result.push(matrix[i as usize][(right - 1) as usize]);
            }
            right -= 1;

            if !(left < right && top < bottom) {
                break;
            }

            for i in (left..right).rev() {
                result.push(matrix[(bottom - 1) as usize][i as usize]);
            }
            bottom -= 1;
            for i in (top..bottom).rev() {
                result.push(matrix[i as usize][left as usize]);
            }
            left += 1;
        }

        result
    }
}
